smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java

```diff
@@ -0,0 +1,364 @@
+     *
+     * @param listener The reactions listener to be removed from the set.
+     * @return {@code true} if the listener was successfully removed;
+     *         {@code false} if the listener was not found in the set.
+     */
+    public synchronized boolean removeReactionsListener(ReactionsListener listener) {
+        return listeners.remove(listener);
+    }
+
+    /**
+     * Creates a form for reaction restrictions, including the max number of reactions per user
+     * and the list of allowed emojis.
+     *
+     * @param maxReactionsPerUser The maximum number of reactions allowed per user.
+     * @param allowedEmojis The list of allowed emojis.
+     * @return The reaction restrictions form.
+     */
+    public DataForm createReactionRestrictionsForm(int maxReactionsPerUser, List<String> allowedEmojis) {
+
+        DataForm.Builder builder = DataForm.builder();
+        builder.setFormType(String.valueOf(DataForm.Type.result));
+
+        builder.addField(
+                        FormField.builder("max_reactions_per_user").setValue(String.valueOf(maxReactionsPerUser))
+                                        .build()
+        );
+
+        FormField.Builder<TextSingleFormField, TextSingleFormField.Builder> allowlistFieldBuilder = FormField.builder("allowlist");
+        for (String emoji : allowedEmojis) {
+            Reaction reaction = new Reaction(emoji);
+            FormField.builder("value").setValue(reaction.getEmoji());
+        }
+        builder.addField(allowlistFieldBuilder.build());
+
+        return builder.build();
+    }
+
+    /**
+     * Advertises reaction restrictions to a given XMPP server.
+     *
+     * @param connection The XMPP connection.
+     * @param maxReactionsPerUser The maximum number of reactions allowed per user.
+     * @param allowedEmojis The list of allowed emojis.
+     */
+    public void advertiseReactionRestrictions(XMPPConnection connection, int maxReactionsPerUser, List<String> allowedEmojis) {
+        ServiceDiscoveryManager sdm = ServiceDiscoveryManager.getInstanceFor(connection);
+        DataForm restrictionsForm = createReactionRestrictionsForm(maxReactionsPerUser, allowedEmojis);
+        sdm.addExtendedInfo(restrictionsForm);
+
+        sdm.addFeature(ReactionsElement.NAMESPACE);
+    }
+
+    /**
+     * Retrieves the reaction restrictions for a given user.
+     *
+     * @param jid The JID of the user.
+     * @return The reaction restrictions for the user.
+     * @throws XMPPException.XMPPErrorException If an XMPP error occurs.
+     * @throws SmackException.NotConnectedException If the connection is not established.
+     * @throws InterruptedException If the operation is interrupted.
+     * @throws SmackException.NoResponseException If no response is received from the server.
+     */
+    public ReactionRestrictions getReactionRestrictions(EntityBareJid jid) throws XMPPException.XMPPErrorException, SmackException.NotConnectedException,
```

Comment:
`EntityBareJId` here is wrong, you want to be able to query all XMPP addresses here (at least, that is my interpretation of the XEP).